### DISCUSSION TITLE
Improved log web view

### DISCUSSION THREAD

**Discussion by ecodina** [apache/airflow#25537] [General]
### Description

The general idea would be to improve the built-in log view inside the webpage. Instead of "just" showing the log file on the browser, the webpage could parse it and help the user find what they are looking for.

I could not find similar requests, either as issues or inside AIP-38.

### Use case/motivation

Some logs are quite long and, since Airflow is quite verbose, it is sometimes difficult to follow all the text. I have seen that Airflow is now working on AIP-38: Modern Web Application, so I think this feature request comes at a good time.

I think that the usability of the webpage could be greatly improved if the following items were implemented:

- Color scheme: different colors for the time, module, type (error, info, debug, ...) and actual text.
- Possibility to filter: only show one type or one module.
- Possibility to search.
- Automatic refresh (not as necessary), so that we can see the new lines on the log file without refreshing the web page

As a good example, I recommend taking a look at how Prefect handles logs.

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
Airflow has already plenty of integration (and you can write more of those) to dedicated systems for logging. https://airflow.apache.org/docs/apache-airflow-providers/core-extensions/logging.html 

There is little value in replicating what those systems are doing - we will never be as good as them in kmplementing all the cool features (including searching, surfacing, monitoring, filtering and what not) as they are. There are deticated teams of people to build such systems, and making them work at scale is a complex things on its own.

Simply Airlfow is a platform and you can extend it's capabilities by connecting to dedicated external systems that do the job better than airflow on particular things it needs:

https://airflow.apache.org/docs/apache-airflow-providers/core-extensions/index.html

And we are even planning to strengthen that by making our logging integrated with Open Telemetry (long term it will also support logging) https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-49+OpenTelemetry+Support+for+Apache+Airflow

I don't think we ever venture in doing anyhing like that. But Let's leave it as a discussion, maybe others will have different opinion. In any case - if there is a different proposal, it should be handled as an AIP not as a feature request as this is quite a change in the general concept of Airlfow as a Platform.

**Reply by ecodina** [reply]
Thanks @potiuk for your detailed answer!

I was aware that there were integrations for logging. However, having everything centralized in one platform - Airflow -, even if it does not have all the cool features these integrations have, makes things easier (I believe).

For example, if we used Elasticsearch with a Kibana frontend, this would mean maintaining 3 different platforms / software. Moreover, whenever a task failed, we would need to log in to Airflow and then to Kibana.

Let's see if anyone else has an opinion.

**Reply by potiuk** [reply]
> I was aware that there were integrations for logging. However, having everything centralized in one platform - Airflow -, even if it does not have all the cool features these integrations have, makes things easier (I believe).

It really depends. There is no "universal" easier or more complex. Everything depends on context. What's easier for you, makes it more complex for those who already use and rely on those external systems to run their business. 

I think what we are really talking about is this dicsussion by @vikramkoka https://lists.apache.org/thread/smfpovyh1hfych9wmbsnvwt397hh7g11 and the way how we segment our users:

From Vikram's email (and I very much agree with it) - we have those 3 groups:

>1. Individuals and small teams using Airflow for their purpose,
> 2. Enterprises managing Airflow for large teams of data engineers and data
scientists, and
> 3. Service providers making "Airflow as a service" available for many
customers, either external or internal.

The "Airlfow as a Platform" approach is more appropriate for groups 2. and 3. where your proposal fits better group 1. So it's not the matter of whether things are easier or more complex, but who the target audience is.

BTW. I am not at all "totally against" making it "nicer" for small users. Quite the opposite. The question is however if  we are not losing opportunities elsewhere and whether we do not complicate too much the basic functionality of Airlfow by implementing such features and whether we are not impairing the "enterprise'y" featuers. 

WDYT @vikramkoka. also @pingzh @ferruzzi I wonder what you think as well since you were participating in the devlist thread ?

**Reply by ferruzzi** [reply]
I mean... I don't see anything against making the logs cleaner. I wouldn't reject a PR that adds auto-refresh (I would actually really like to see that one added, personally), or to add colors to the logs, but I don't see adding features like search/filter as a top priority, personally.

**Comment by jedcunningham** [comment]
We actually have a bit of this already coming in 2.4.0, thanks to @pierrejeambrun in #24403. For example:

![task log filtering](https://user-images.githubusercontent.com/66968678/183539818-2e9d6cba-2b9d-4b0e-8cce-0d41e9e8dc07.png)

**Reply by potiuk** [reply]
Cool. If we can make something that nice and simple :)

**Reply by ecodina** [reply]
This looks great and it will be of great help! Thank you!

**Comment by pierrejeambrun** [comment]
As pointed by @jedcunningham some of this will be address in 2.4.

I agree that trying to be as good as specialized logging solution brings little value to airflow especially when we can easily integrate with these tools.

Logs auto refresh is nice :tada: We wanted to add it to the dedicated 'logs' page but it required some effort (FAB). Now that logs are in the 'React' world, it is really simple to add this feature. I just opened a PR for this:
https://github.com/apache/airflow/pull/25621

As usual you can configure the refetch interval via the `auto_refresh_interval` param. (3s by default)

**Comment by potiuk** [comment]
I personally think what I see in those PRs is perfect - nicely joining beauty/modernising the log experience without spending awfully lot of time on much more sophisticated experience (where dedicated logging solutions will always be ahead of us).

**Comment by potiuk** [comment]
And I really love how the new modern react framework makes it easy to do such things like autorefresh
